src/runtime/profiler.cpp

```diff
@@ -377,6 +662,66 @@ WEAK void halide_profiler_report_unlocked(void *user_context, halide_profiler_st
+                        const double cpu_cycles = extrapolate_counter(h + halide_profiler_hardware_counter_cpu_cycles);
+                        const uint64_t ipc_times_100 = (uint64_t)((100.0 * instructions) / (cpu_cycles + 1e-20) + 0.5);
+                        const uint64_t ipc_hundreds = ipc_times_100 / 100;
+                        const uint64_t ipc_frac = ipc_times_100 - ipc_hundreds*100;
+                        const uint64_t ipc_tens = ipc_frac / 10;
+                        const uint64_t ipc_ones = ipc_frac % 10;
+                        const double l1_accesses = extrapolate_counter(h + halide_profiler_hardware_counter_l1d_read_accesses);
+                        const double l1_misses = extrapolate_counter(h + halide_profiler_hardware_counter_l1d_read_misses);
+                        const uint64_t l1_hit_rate = 100 - (uint64_t)((100.0 * l1_misses) / (l1_accesses + 1e-20) + 0.5);
+                        const double llc_accesses = extrapolate_counter(h + halide_profiler_hardware_counter_llc_read_accesses);
+                        const double llc_misses = extrapolate_counter(h + halide_profiler_hardware_counter_llc_read_misses);
+                        const uint64_t llc_hit_rate = 100 - (uint64_t)((100.0 * llc_misses) / (llc_accesses + 1e-20) + 0.5);
+
+                        double l1_bytes = l1_accesses * 64.0;
+                        double llc_bytes = llc_accesses * 64.0;
+
+                        l1_bytes /= p->runs;
+                        llc_bytes /= p->runs;
+
+                        const char *suffixes[] = {"B", "K", "M", "G", "T", NULL};
+                        int l1_suffix = 0, llc_suffix = 0;
+                        while (l1_bytes >= 10000 && suffixes[l1_suffix + 1]) {
+                            l1_suffix ++;
+                            l1_bytes /= 1000;
+                        }
+                        while (llc_bytes >= 10000 && suffixes[llc_suffix + 1]) {
+                            llc_suffix ++;
+                            llc_bytes /= 1000;
+                        }
+
+                        sstr << "                      "
+                             << "   IPC: " << ipc_hundreds << "." << ipc_tens << ipc_ones
+                             << "   L1: " << (uint64_t)(l1_bytes + 0.5) << suffixes[l1_suffix] << " (" << l1_hit_rate << "%)"
+                             << "   LLC: " << (uint64_t)(llc_bytes + 0.5) << suffixes[llc_suffix] << " (" << llc_hit_rate << "%)"
+                             << "\n";
+                        halide_print(user_context, sstr.str());
+                    } else {
+                        // Just print the counter values
+                        uint64_t a = s->active_counters;
+                        halide_profiler_hardware_counter *c = (halide_profiler_hardware_counter *)(fs->hardware_counters);
+                        while (a) {
+                            uint64_t i = __builtin_ctzl(a);
```

Comment:
Same comment as on the other instance of this. Especially here because performance is less of a concern (but I can see the value of having the logic be the same in both places).